ethereum/blockwatch/block_watcher_test.go

```diff
@@ -67,25 +72,398 @@ func TestWatcher(t *testing.T) {
+			rangeToFilterLogsResponse: map[string]filterLogsResponse{
+				"10-20": filterLogsResponse{
+					Err: errors.New(infuraTooManyResultsErrMsg),
+				},
+				"10-15": filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+				"16-20": filterLogsResponse{
+					Err: errors.New(infuraTooManyResultsErrMsg),
+				},
+				"16-18": filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+				"19-20": filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+			},
+			Logs: []types.Log{logStub, logStub, logStub},
+		},
+		filterLogsRecusivelyTestCase{
+			Label: "TOO_MANY_RESULTS_INFURA_ERROR_DEEPER_RECURSION_FAILURE",
+			rangeToFilterLogsResponse: map[string]filterLogsResponse{
+				"10-20": filterLogsResponse{
+					Err: errors.New(infuraTooManyResultsErrMsg),
+				},
+				"10-15": filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+				"16-20": filterLogsResponse{
+					Err: errUnexpected,
+				},
+			},
+			Err: errUnexpected,
+		},
+		filterLogsRecusivelyTestCase{
+			Label: "UNEXPECTED_ERROR",
+			rangeToFilterLogsResponse: map[string]filterLogsResponse{
+				"10-20": filterLogsResponse{
+					Err: errUnexpected,
+				},
+			},
+			Err: errUnexpected,
+		},
+	}
+
+	meshDB, err := meshdb.NewMeshDB("/tmp/leveldb_testing/" + uuid.New().String())
+	require.NoError(t, err)
+	config.MeshDB = meshDB
+
+	for _, testCase := range testCases {
+		fakeLogClient, err := newFakeLogClient(testCase.rangeToFilterLogsResponse)
+		require.NoError(t, err)
+		config.Client = fakeLogClient
+		watcher := New(config)
+
+		logs, err := watcher.filterLogsRecurisively(from, to, []types.Log{})
+		require.Equal(t, testCase.Err, err, testCase.Label)
+		require.Equal(t, testCase.Logs, logs, testCase.Label)
+		assert.Equal(t, len(testCase.rangeToFilterLogsResponse), fakeLogClient.Count())
+	}
+}
+
+type logsInBlockRangeTestCase struct {
+	Label                     string
+	From                      int
+	To                        int
+	RangeToFilterLogsResponse map[string]filterLogsResponse
+	Logs                      []types.Log
+	FurthestBlockProcessed    int64
+}
+
+func TestGetLogsInBlockRange(t *testing.T) {
+	from := 10
+	to := 20
+	maxBlocksInQuery := int(maxBlocksInGetLogsQuery)
+	testCases := []logsInBlockRangeTestCase{
+		logsInBlockRangeTestCase{
+			Label: "HAPPY_PATH",
+			From:  from,
+			To:    to,
+			RangeToFilterLogsResponse: map[string]filterLogsResponse{
+				toR(from, to): filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+			},
+			Logs:                   []types.Log{logStub},
+			FurthestBlockProcessed: int64(to),
+		},
+		logsInBlockRangeTestCase{
+			Label: "SPLIT_REQUEST_BY_MAX_BLOCKS_IN_QUERY",
+			From:  from,
+			To:    from + maxBlocksInQuery + 10,
+			RangeToFilterLogsResponse: map[string]filterLogsResponse{
+				toR(from, from+maxBlocksInQuery-1): filterLogsResponse{
```

Comment:
We should consider using constants in the range here just to make sure that the result is exactly what we expect. Using arithmetic runs the risk of accidentally introducing a bug in the test cases.

Reply:
Although I generally agree, I feel inclined to keep these tests this way. I feel it's much clearer _what_ we are testing by keeping these as simple arithmetic of well-named constants. Much more so then e.g., `"10-71"`

Reply:
Fair enough. I'm okay with it.